### PR TITLE
Load allow vertical scale flag in reconciler manager

### PR DESCRIPTION
**Pull request by victorpras**
Load --allow-vertical-scale flag from reconcilerManager deployment, keep current container resources when flag is enabled, and add e2e test.

--allow-vertical-scale flag is declared in the reconcilerManager deployment when allowVerticalScale is set to true in the ConfigManagement API. When the flag is enabled, the reconciler should load the flag, and copy over all containers' resources from the current reconciler Deployment to the declared manifest, so that the discrepancy will not cause any Deployment update. That implies any resource override applied to the declared Deployment will be ignored. Changes to anything other than the resource will not be maintained and will be reverted back

The e2e test will simulate the scenario by adding the --allow-vertical-scale flag to the reconciler-manager deployment, perform an update to the resource of a container in the reconciler deployment and another update to minReadySeconds field. The update to the resource should then be maintained and not reverted, while minReadySeconds update is reverted

### CODE REVIEW

File: e2e/testcases/reconciler_manager_test.go [hunk 1]
```diff
@@ -603,3 +603,44 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	firstContainerName := reconcilerDeployment.Spec.Template.Spec.Containers[0].Name
+	updatedFirstContainerCPURequest = 159
+	updatedFirstContainerMemoryRequest = 259
+
+	if allowVerticalScale {
+		nt.MustMergePatch(rs, fmt.Sprintf(`{"spec":{"override":{"resources":[{"containerName":"%s","memoryRequest":"%dMi", "cpuRequest":"%dm"}]}}}`,
+			firstContainerName, updatedFirstContainerMemoryRequest, updatedFirstContainerCPURequest))
+	}
+
+	nomostest.Wait(nt.T, "Waiting for fourth reconciliation", nt.DefaultWaitTimeout, func() error {
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:641]: Update the description

File: e2e/testcases/reconciler_manager_test.go [hunk 2]
```diff
@@ -603,3 +603,44 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+		return nt.Validate(nomostest.DefaultRootReconcilerName, v1.NSConfigManagementSystem, &appsv1.Deployment{},
+			hasGeneration(3), firstContainerCPURequestIs(updatedFirstContainerCPURequest),
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:643]: Add a comment to explain why generation is 3.

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 3]
```diff
@@ -172,6 +172,15 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
 
 	// TODO: check if VPA is enabled
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/reconciler_base.go:173]: Remove the TODO comment since this commit resolves it.

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 4]
```diff
@@ -172,6 +172,15 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
...
 	vpaEnabled := false
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/reconciler_base.go:174]: Rename the variable to avoid confusing.

File: e2e/testcases/reconciler_manager_test.go [hunk 5]
```diff
@@ -603,3 +603,44 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	allowVerticalScale := false
+	for _, c := range reconcilerDeployment.Spec.Template.Spec.Containers {
+		if c.Name == "reconciler-manager" {
+			for _, arg := range c.Args {
+				if arg == "--allow-vertical-scale" {
+					allowVerticalScale = true
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:622]: enabling the flag process has been changed to use patch

File: e2e/testcases/reconciler_manager_test.go (hunk 1 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:633]: description udpated

File: e2e/testcases/reconciler_manager_test.go [hunk 6]
```diff
@@ -603,3 +603,37 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	var updatedFirstContainerMemoryRequest int64
+	firstContainerName := reconcilerDeployment.Spec.Template.Spec.Containers[0].Name
+	updatedFirstContainerCPURequest = 159
+	updatedFirstContainerMemoryRequest = 259
+
+	nt.MustMergePatch(rs, fmt.Sprintf(`{"spec":{"override":{"resources":[{"containerName":"%s","memoryRequest":"%dMi", "cpuRequest":"%dm"}]}}}`,
+		firstContainerName, updatedFirstContainerMemoryRequest, updatedFirstContainerCPURequest))
+
+	// generation = 2 because the applied changes will not be reverted
+	nomostest.Wait(nt.T, "Waiting for fourth reconciliation", nt.DefaultWaitTimeout, func() error {
```
Comment (janetkuo) [inline e2e/testcases/reconciler_manager_test.go:634]: Why is this the fourth reconciliation?

File: e2e/testcases/reconciler_manager_test.go [hunk 7]
```diff
@@ -603,3 +603,37 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	_, err := nt.Kubectl("patch", "deployment", "reconciler-manager", "-n", "config-management-system", "-p", "{\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"args\":[\"--enable-leader-election\", \"--allow-vertical-scale\"],\"name\":\"reconciler-manager\"}]}}}}")
+	if err != nil {
+		nt.T.Fatalf("Error enabling vertical scale in deployment", err)
+	}
+
+	var updatedFirstContainerCPURequest int64
+	var updatedFirstContainerMemoryRequest int64
+	firstContainerName := reconcilerDeployment.Spec.Template.Spec.Containers[0].Name
+	updatedFirstContainerCPURequest = 159
+	updatedFirstContainerMemoryRequest = 259
```
Comment (janetkuo) [inline e2e/testcases/reconciler_manager_test.go:628]: Would this pass in Autopilot clusters?

Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:628]: the generation count could be different in autopilot cluster, i will add a condition to skip to separate the autopilot validation

**Comment by sdowell** [comment]
I suggest rebasing instead of adding merge commits, I don't think it will let you submit with merge commits. The PR commit history is also easier to follow with rebasing

File: e2e/testcases/reconciler_manager_test.go (hunk 6 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: Updated logs to be more clear

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 4 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go]: variable has been renamed

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 3 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go]: replaced todo with clarifying comments

File: e2e/testcases/reconciler_manager_test.go (hunk 2 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: generation should be 2, added explanation comment

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 8]
```diff
@@ -253,16 +266,21 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
...
+//
 // 3. If VPA is not enabled, and it is an Autopilot cluster, the controller will
-//    honor the resource override, but update the declared resources to be compliant
-//    with the Autopilot resource range constraints.
-func adjustContainerResources(vpaEnabled, isAutopilot bool, declared, current *appsv1.Deployment) (bool, error) {
+//
+//	honor the resource override, but update the declared resources to be compliant
+//	with the Autopilot resource range constraints.
+func adjustContainerResources(allowVerticalScale, isAutopilot bool, declared, current *appsv1.Deployment) (bool, error) {
 	// If VPA is enabled, use the current resources because VPA takes full control of them.
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reconciler_base.go:282]: Update comment

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 9]
```diff
@@ -253,16 +266,21 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
 // This function aims to address the fight among Autopilot, VPA and the reconciler since they all update the resources.
 // Below is the resolution:
 // 1. If VPA is enabled (no matter if it is a standard cluster or an Autopilot cluster),
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reconciler_base.go:268]: Need to update all comments about "VPA enabled". We didn't check whether VPA is enabled in the cluster or not. We're only checking whether the --allow-vertical-scale flag is enabled in the reconciler manager.

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 10]
```diff
@@ -253,16 +266,21 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
...
-//    use the current resources and ignore the override defined in the declared Deployment
-//    because VPA will scale up and down automatically.
+//
+//	use the current resources and ignore the override defined in the declared Deployment
+//	because VPA will scale up and down automatically.
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reconciler_base.go:271]: It's not VPA specific. We can remove L271.

File: cmd/reconciler-manager/main.go [hunk 11]
```diff
@@ -70,10 +70,12 @@ func init() {
 func main() {
 	var metricsAddr string
 	var enableLeaderElection bool
+	var allowVerticalScale bool
 	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&allowVerticalScale, "allow-vertical-scale", false, "Allowing vertical scale to let VPA manage the resource")
```
Comment (janetkuo) [inline cmd/reconciler-manager/main.go:78]: `allowVerticalScale` is never used.

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 12]
```diff
@@ -194,8 +194,21 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
 		return controllerutil.OperationResultCreated, nil
 	}
 
-	// TODO: check if VPA is enabled
-	vpaEnabled := false
+	//check if allow vertical scale is enabled
+	allowVerticalScale := false
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reconciler_base.go:198]: You should be able to get the setting from the reconciler flag var `allowVerticalScale` and pass it here, instead of making an API to get the information from the cluster.

File: e2e/testcases/reconciler_manager_test.go (hunk 6 above)
Comment (janetkuo) [reply e2e/testcases/reconciler_manager_test.go:633]: I didn't see your update. Did you push the change?

File: e2e/testcases/reconciler_manager_test.go (hunk 7 above)
Comment (janetkuo) [reply e2e/testcases/reconciler_manager_test.go:628]: Do we need to skip in Autopilot? This should still work in Autopilot cluster. Maybe you can check whether a the CPU/memory request is changed by Autopilot, or just make your request compliant with Autopilot.

File: e2e/testcases/reconciler_manager_test.go [hunk 13]
```diff
@@ -603,3 +603,36 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+
+	_, err := nt.Kubectl("patch", "deployment", "reconciler-manager", "-n", "config-management-system", "-p", "{\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"args\":[\"--enable-leader-election\", \"--allow-vertical-scale\"],\"name\":\"reconciler-manager\"}]}}}}")
+	if err != nil {
+		nt.T.Fatalf("Error enabling vertical scale in deployment", err)
+	}
+
+	var updatedFirstContainerCPURequest int64
+	var updatedFirstContainerMemoryRequest int64
+	firstContainerName := reconcilerDeployment.Spec.Template.Spec.Containers[0].Name
+	updatedFirstContainerCPURequest = 159
```
Comment (janetkuo) [inline e2e/testcases/reconciler_manager_test.go:627]: Need to test limits as well.

File: e2e/testcases/reconciler_manager_test.go (hunk 6 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:633]: sorry, looks like i lost a few commits when squashing them. Fixing now

File: cmd/reconciler-manager/main.go (hunk 11 above)
Comment (victorpras) [reply cmd/reconciler-manager/main.go:78]: allowVerticalScale will be propagated from reconcilerBase

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 9 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go:268]: update comments

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 12 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go:198]: as per previous comment, allowVerticalScale will be passed using reconcilerBase

File: pkg/reconcilermanager/controllers/reconciler_base.go [hunk 14]
```diff
@@ -194,8 +195,8 @@ func (r *reconcilerBase) createOrPatchDeployment(ctx context.Context, declared *
 		return controllerutil.OperationResultCreated, nil
 	}
 
-	// TODO: check if VPA is enabled
-	vpaEnabled := false
+	//check if allow vertical scale is enabled
+	allowVerticalScale := r.allowVerticalScale
```
Comment (janetkuo) [inline pkg/reconcilermanager/controllers/reconciler_base.go:199]: You can just use `r.allowVerticalScale` directly

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 10 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go:258]: comments have been updated to a more clear description

File: e2e/testcases/reconciler_manager_test.go (hunk 7 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:630]: added separate autopilot and regular cluster validation

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 8 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go]: comment updated

File: pkg/reconcilermanager/controllers/reconciler_base.go (hunk 14 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reconciler_base.go]: used r.allowVerticalScale directly

**Comment by victorpras** [comment]
All comments are addressed. PTAL

File: e2e/testcases/reconciler_manager_test.go (hunk 13 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: limit verification has been added

File: cmd/reconciler-manager/main.go (hunk 11 above)
Comment (nan-yu) [inline cmd/reconciler-manager/main.go:78]: Can you define an environment variable and load the default value from the env variable? This provides more flexibility for OSS users to set the flag.

File: pkg/reconcilermanager/controllers/reposync_controller_test.go [hunk 15]
```diff
@@ -281,6 +281,7 @@ func setupNSReconciler(t *testing.T, objs ...client.Object) (*syncerFake.Client,
 		fakeClient,
 		controllerruntime.Log.WithName("controllers").WithName("RepoSync"),
 		s,
+		false,
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/reposync_controller_test.go:284]: Can you add unit test when allowVerticalScale is set to true?

File: e2e/testcases/reconciler_manager_test.go [hunk 16]
```diff
@@ -603,3 +603,48 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	var updatedFirstContainerCPURequest int64
+	var updatedFirstContainerMemoryRequest int64
+	var autopilotFirstContainerCPURequest int64
+
+	firstContainerName := reconcilerDeployment.Spec.Template.Spec.Containers[0].Name
+	updatedFirstContainerCPURequest = 249
+	updatedFirstContainerMemoryRequest = 259
+	autopilotFirstContainerCPURequest = 430
+
+	nt.MustMergePatch(rs, fmt.Sprintf(`{"spec":{"override":{"resources":[{"containerName":"%s","memoryRequest":"%dMi", "cpuRequest":"%dm"}]}}}`,
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:633]: I suppose this is to simulate the VPA behavior? However, VPA won't update the override value in the RSync API. Instead, it updates the reconciler Deployment directly.

When you fix it, please add a comment to explain this is to simulate the VPA behavior.

File: e2e/testcases/reconciler_manager_test.go [hunk 17]
```diff
@@ -603,3 +603,48 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+		firstContainerName, updatedFirstContainerMemoryRequest, updatedFirstContainerCPURequest))
+
+	if !nt.IsGKEAutopilot {
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:636]: In the implementation, you updated a comment to explain that when VPA is enabled, the behavior should be the same in standard cluster and GKE. Please update the test to reflect what you implemented.

File: e2e/testcases/reconciler_manager_test.go (hunk 16 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: I updated the test to use ` kubectl patch ` to directly update the reconciler deployment to simulate actual VPA behavior.

File: e2e/testcases/reconciler_manager_test.go (hunk 17 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: validation has been updated to reflect the implementation for standard and autopilot cluster

File: pkg/reconcilermanager/controllers/reposync_controller_test.go (hunk 15 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reposync_controller_test.go]: unit test has been added. The unit test will 
1. Add allowVerticalScale to test reconciler to mock real allowVerticalScale setting
2. Update the resource and replica directly in the fakeClient
3. Reconcile
4. Validate that resource update are maintained, but replica update is reverted

File: pkg/reconcilermanager/controllers/reposync_controller_test.go [hunk 18]
```diff
@@ -2532,6 +2534,67 @@ func TestRepoSyncReconcileStaleClientCache(t *testing.T) {
...
+	}
+
+	// when vertical scale is enabled, original deployment resource should not be updated
+	deploymentCoreObject = fakeClient.Objects[core.IDOf(repoDeployment)]
+	wantDeployments = deploymentCoreObject.(*appsv1.Deployment)
+	for _, container := range wantDeployments.Spec.Template.Spec.Containers {
+		if container.Name == reconcilermanager.Reconciler {
+			currentCPU := container.Resources.Requests.Cpu().String()
+			if updatedCPU != currentCPU {
+				t.Errorf("Deployment validation failed. Expected CPU: %s, got %s", updatedCPU, currentCPU)
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/reposync_controller_test.go:2584]: Please reuse the `validateDeployments` function to verify.

File: pkg/reconcilermanager/controllers/rootsync_controller_test.go [hunk 19]
```diff
@@ -136,6 +136,7 @@ func setupRootReconciler(t *testing.T, objs ...client.Object) (*syncerFake.Clien
 		fakeClient,
 		controllerruntime.Log.WithName("controllers").WithName("RootSync"),
 		s,
+		false,
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/rootsync_controller_test.go:139]: Add unit test for the rootsync controller as well.

File: pkg/reconcilermanager/controllers/reposync_controller_test.go (hunk 18 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/reposync_controller_test.go]: validation has been modified to construct expected deployment and using validateDeployments function

File: e2e/testcases/reconciler_manager_test.go [hunk 20]
```diff
@@ -603,3 +603,40 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	reconcilerDeployment := &appsv1.Deployment{}
+	if err := nt.Validate(nomostest.DefaultRootReconcilerName, v1.NSConfigManagementSystem, reconcilerDeployment); err != nil {
+		nt.T.Fatal(err)
+	}
+	rs := &v1beta1.RootSync{}
+	if err := nt.Get(configsync.RootSyncName, configsync.ControllerNamespace, rs); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	// enable vertical scale
```
[the remark this answers is not in the document]
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: fixed typo

File: cmd/reconciler-manager/main.go [hunk 21]
```diff
@@ -70,10 +73,13 @@ func init() {
 func main() {
 	var metricsAddr string
 	var enableLeaderElection bool
+	var allowVerticalScale bool
 	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	defaultAllowVerticalScale, _ := strconv.ParseBool(verticalScale)
+	flag.BoolVar(&allowVerticalScale, "allow-vertical-scale", defaultAllowVerticalScale, "Allowing vertical scale to let VPA manage the resource")
```
[the remark this answers is not in the document]
Comment (victorpras) [reply cmd/reconciler-manager/main.go]: thanks for the suggestion! refactored to use EnvBool

File: pkg/reconcilermanager/controllers/rootsync_controller_test.go (hunk 19 above)
Comment (victorpras) [reply pkg/reconcilermanager/controllers/rootsync_controller_test.go]: Root sync controller test has been added, follows the same structure with the repo sync test

File: cmd/reconciler-manager/main.go (hunk 11 above)
Comment (victorpras) [reply cmd/reconciler-manager/main.go]: As per other comment, this has been updated to use EnvBool

File: e2e/testcases/reconciler_manager_test.go [hunk 22]
```diff
@@ -603,3 +603,38 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+	var updatedFirstContainerCPURequest int64 = 430
+	var updatedFirstContainerMemoryRequest int64 = 200
+
+	// simulating VPA update to deployment
+	_, err = nt.Kubectl("patch", "deployment", "root-reconciler", "-n", "config-management-system", "-p", fmt.Sprintf("{\"spec\":{\"template\":{\"spec\":{\"containers\":[{\"name\":\"%s\",\"resources\":{\"requests\":{\"cpu\":\"%dm\",\"memory\":\"%dMi\"}}}]}}}}", "hydration-controller", updatedFirstContainerCPURequest, updatedFirstContainerMemoryRequest))
+	if err != nil {
+		nt.T.Fatalf("Error simulating VPA deployment update", err)
+	}
+
+	nomostest.Wait(nt.T, "Waiting for patch deployment update", nt.DefaultWaitTimeout, func() error {
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:635]: This is not waiting for patch deployment update. This is to confirm the reconciler-manager doesn't revert the change.
I would suggest you update another field of the Deployment, and validate that field can be reverted by the reconciler-manager, while the "VPA adjusted resources" remain the same.

File: pkg/reconcilermanager/controllers/reposync_controller_test.go [hunk 23]
```diff
@@ -2585,6 +2587,72 @@ func TestRepoSyncReconcileStaleClientCache(t *testing.T) {
...
+	if err := fakeClient.Update(ctx, updatedDeployment); err != nil {
+		t.Fatalf("failed to update the repo sync request, got error: %v", err)
+	}
+	if _, err := testReconciler.Reconcile(ctx, reqNamespacedName); err != nil {
+		t.Fatalf("unexpected reconciliation error upon request update, got error: %q, want error: nil", err)
+	}
+
+	// when allow vertical scale is enabled, original deployment resource should not be updated
+	// other updated deployment should be reverted
+	overrideResource := []v1beta1.ContainerResourcesSpec{
```
Comment (nan-yu) [inline pkg/reconcilermanager/controllers/reposync_controller_test.go:2632]: Rename this variable because it is not the override.

Comment (victorpras) [reply pkg/reconcilermanager/controllers/reposync_controller_test.go]: renamed for clarity

File: e2e/testcases/reconciler_manager_test.go (hunk 22 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: updated comments for clarity.

added minReadySeconds update to the test and verify it getting reverted

File: e2e/testcases/reconciler_manager_test.go [hunk 24]
```diff
@@ -603,3 +603,38 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+		return nt.Validate("root-reconciler", v1.NSConfigManagementSystem, &appsv1.Deployment{},
+			hasGeneration(2), firstContainerCPURequestIs(updatedFirstContainerCPURequest),
```
[the remark this answers is not in the document]
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: in the previous commit, the test was updating `reconciler-manager` deployment. Since there was a --enable-vertical-scale flag addition (+1) and container resource update (+1), the expected generation is 3. The test has been modified to update `hydration-controller` container in `root-reconciler` deployment instead, so there's only resource update (+1) therefore the expected deployment is 2. After this feedback, we add `minReadySeconds` update (+1) that should be reverted (+1), so the expected deployment is now 3

File: e2e/testcases/reconciler_manager_test.go [hunk 25]
```diff
@@ -603,3 +616,39 @@ func TestAutopilotReconcilerAdjustment(t *testing.T) {
...
+		nt.T.Fatalf("Error enabling vertical scale in deployment", err)
+	}
+
+	// update numbers are compliant with autopilot requirement
+	var updatedFirstContainerCPURequest int64 = 430
+	var updatedFirstContainerMemoryRequest int64 = 200
+	initialMinReadySeconds := reconcilerDeployment.Spec.MinReadySeconds
+
+	// simulating VPA update to deployment
+	_, err = nt.Kubectl("patch", "deployment", "root-reconciler", "-n", "config-management-system", "-p", fmt.Sprintf("{\"spec\":{\"minReadySeconds\":%d,\"template\":{\"spec\":{\"containers\":[{\"name\":\"%s\",\"resources\":{\"requests\":{\"cpu\":\"%dm\",\"memory\":\"%dMi\"}}}]}}}}", 25, "hydration-controller", updatedFirstContainerCPURequest, updatedFirstContainerMemoryRequest))
```
Comment (nan-yu) [inline e2e/testcases/reconciler_manager_test.go:644]: nit: this is a constant. You can use it in the format directly `\"minReadySeconds\":25`.

File: e2e/testcases/reconciler_manager_test.go (hunk 2 above)
Comment (nan-yu) [reply e2e/testcases/reconciler_manager_test.go:651]: [-2] Is generation 2 or 3? It is changed to 3 in your latest commit now. I didn't see explanation comment either.

File: e2e/testcases/reconciler_manager_test.go (hunk 25 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: fixed

File: e2e/testcases/reconciler_manager_test.go (hunk 2 above)
Comment (nan-yu) [reply e2e/testcases/reconciler_manager_test.go:652]: Why is this comment marked as resolved? What is the expected generation?

Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:654]: as explained in https://github.com/GoogleContainerTools/kpt-config-sync/pull/20#discussion_r956371505, the test works differently in the earlier commits, therefore the expected generation was different

File: e2e/testcases/reconciler_manager_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go]: unresolving the comments for now, let me know if this makes sense

File: e2e/testcases/reconciler_manager_test.go (hunk 2 above)
Comment (victorpras) [reply e2e/testcases/reconciler_manager_test.go:654]: unresolving for now since this is related with https://github.com/GoogleContainerTools/kpt-config-sync/pull/20#discussion_r956371505

File: e2e/testcases/reconciler_manager_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (nan-yu) [reply e2e/testcases/reconciler_manager_test.go]: ack

File: e2e/testcases/reconciler_manager_test.go (hunk 2 above)
Comment (nan-yu) [reply e2e/testcases/reconciler_manager_test.go:654]: ack

**Review by nan-yu — COMMENTED** [review]
Looks good. Can you squash your commits into a single commit?

**Comment by victorpras** [comment]
All commits have been squashed to a single commit

**Review by nan-yu — APPROVED** [review]
/lgtm
